src/web/http/connection.cpp

```diff
@@ -0,0 +1,430 @@
+}
+
+buffer& connection::read_buffer()
+{
+    return read_buffer_;
+}
+
+data_buffer& connection::write_buffer()
+{
+    return write_buffer_;
+}
+
+// This is effectively a blocking write call that does not buffer
+// internally.
+int32_t connection::do_write(const unsigned char* data, int32_t length,
+    bool write_frame)
+{
+    write_method plaintext_write = [this](const unsigned char* data,
+        int32_t length)
+    {
+        return static_cast<int32_t>(send(socket_,
+#ifdef WIN32
+            reinterpret_cast<char*>(data),
+#else
+            data,
+#endif
+            length, 0));
+    };
+
+#ifdef WITH_MBEDTLS
+    write_method ssl_write = [this](const unsigned char* data, int32_t length)
+    {
+        return static_cast<int32_t>(
+            mbedtls_ssl_write(&ssl_context_.context, data, length));
+    };
+
+    auto writer = static_cast<write_method>(ssl_context_.enabled ? ssl_write :
+        plaintext_write);
+#else
+    auto writer = static_cast<write_method>(plaintext_write);
+#endif
+
+    if (write_frame)
+    {
+        const auto frame = generate_websocket_frame(length,
+            websocket_op::text);
+        const auto frame_data = reinterpret_cast<const unsigned char*>(&frame);
+
+        int32_t frame_write = writer(frame_data, frame.write_length);
+        if (frame_write < 0)
+            return frame_write;
+    }
+
+    int32_t written = 0;
+    auto remaining = length;
+    auto position = data;
+
+    do
+    {
+        written = writer(position, remaining);
+        if (written < 0)
+        {
+            const auto error = last_error();
+            if (!would_block(error))
+            {
+                LOG_WARNING(LOG_SERVER_HTTP)
+                    << "do_write failed. requested " << remaining
+                    << " and wrote " << written << ": " << error_string();
+                return written;
+            }
+
+            continue;
+        }
+
+        position += written;
+        remaining -= written;
+
+    } while (remaining != 0);
+
+    return static_cast<int32_t>(position - data);
+}
+
+int32_t connection::write(const std::string& buffer)
+{
+    return write(reinterpret_cast<const unsigned char*>(buffer.c_str()),
+        buffer.size());
+}
+
+// This is a buffered write call so long as we're under the high
+// water mark.
+int32_t connection::write(const unsigned char* data, int32_t length)
+{
+    if (length < 0)
+        return length;
+
+    const int32_t buffered_length = write_buffer_.size() + length +
+        (websocket_ ? sizeof(websocket_frame) : 0);
+
+    // If we're currently at the hwm, issue blocking writes until
+    // we've cleared the buffered data and then write this current
+    // request.  This is an expensive operation, but should be
+    // mostly avoidable with proper hwm tuning of your
+    // application.
+    if (buffered_length >= high_water_mark_)
+    {
+        // Drain the buffered data.
+        while (!write_buffer_.empty())
```

Comment:
This throttles the server in order to satisfy a slow client. When high water is hit for a client new to-be-buffered messages should be dropped.